client/logmon/logmon.go

```diff
@@ -197,37 +199,47 @@ func (l *logRotatorWrapper) isRunning() bool {
 // processOutWriter to attach to the stdout or stderr of a process.
 func newLogRotatorWrapper(path string, logger hclog.Logger, rotator *logging.FileRotator) (*logRotatorWrapper, error) {
 	logger.Info("opening fifo", "path", path)
-	f, err := fifo.New(path)
+	fifoOpenFn, err := fifo.New(path)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create fifo for extracting logs: %v", err)
 	}
 
 	wrap := &logRotatorWrapper{
 		fifoPath:          path,
-		processOutReader:  f,
 		rotatorWriter:     rotator,
 		hasFinishedCopied: make(chan struct{}),
+		opened:            make(chan struct{}),
 		logger:            logger,
 	}
-	wrap.start()
+	wrap.start(fifoOpenFn)
 	return wrap, nil
 }
 
 // start starts a goroutine that copies from the pipe into the rotator. This is
 // called by the constructor and not the user of the wrapper.
-func (l *logRotatorWrapper) start() {
+func (l *logRotatorWrapper) start(readerOpenFn func() (io.ReadCloser, error)) {
 	go func() {
 		defer close(l.hasFinishedCopied)
-		_, err := io.Copy(l.rotatorWriter, l.processOutReader)
+
+		reader, err := readerOpenFn()
+		if err != nil {
+			l.logger.Warn("failed to open log fifo", "error", err)
+			return
+		}
+		l.processOutReader = reader
```

Comment:
Do logmon tests pass with `-race` enabled? AFAICT this is safe due to the only other use waiting on `l.opened`, but I just want to make sure as in the old code this field was set statically on struct creation.

Reply:
Yes, I made further tweaks so we detect when open call fails as well - and ensured that -race is still good:

```
vagrant@linux:~/go/src/github.com/hashicorp/nomad$ make dev && sudo -E PATH="${PATH}" gotestsum -- -race github.com/hashicorp/nomad/client/logmon
--> Formatting vendor/vendor.json
test -x /home/vagrant/go/bin/vendorfmt || go get -u github.com/magiconair/vendorfmt/cmd/vendorfmt
vendorfmt
Formatted vendor/vendor.json
--> Making [GH-xxxx] references clickable...
==> Removing old development build...
==> Building pkg/linux_amd64/nomad with tags ...
✓  client/logmon (9.436s)

DONE 2 tests in 21.665s
```